src/litserve/server.py

```diff
@@ -19,15 +20,59 @@
 LIT_SERVER_API_KEY = os.environ.get("LIT_SERVER_API_KEY")
 
 
-def inference_worker(lit_api, device, worker_id, request_queue, request_buffer):
-    lit_api.setup(device=device)
+def get_batch_from_uid(uids, lit_api, request_buffer):
+    batches = []
+    for uid in uids:
+        try:
+            x_enc, pipe_s = request_buffer.pop(uid)
+        except KeyError:
+            continue
+        x = lit_api.decode_request(x_enc)
+        batches.append((x, pipe_s))
+    return batches
+
+
+def collate_requests(lit_api, request_queue: Queue, request_buffer, max_batch_size, batch_timeout):
+    uids = []
+    entered_at = time.time()
+    while (batch_timeout - (time.time() - entered_at) > 0) and len(uids) <= max_batch_size:
+        try:
+            uid = request_queue.get_nowait()
+            uids.append(uid)
+        except (Empty, ValueError):
+            continue
+    return get_batch_from_uid(uids, lit_api, request_buffer)
+
 
+def run_batched_loop(lit_api, request_queue: Queue, request_buffer, max_batch_size, batch_timeout):
+    while True:
+        batches = collate_requests(
+            lit_api,
+            request_queue,
+            request_buffer,
+            max_batch_size,
+            batch_timeout,
+        )
+        if not batches:
+            continue
+
+        inputs, pipes = zip(*batches)
+
+        x = lit_api.batch(inputs)
+        y = lit_api.predict(x)
+        outputs = lit_api.unbatch(y)
+
+        for y, pipe_s in zip(outputs, pipes):
+            y_enc = lit_api.encode_response(y)
+
+            with contextlib.suppress(BrokenPipeError):
+                pipe_s.send(y_enc)
+
+
+def run_single_loop(lit_api, request_queue: Queue, request_buffer):
     while True:
-        # NOTE: to implement batching here: keep getting items from the queue,
-        #       fill a batch, predict, send outputs to the respective pipes
-        #       In the future we will expose this through the API.
         try:
-            uid = request_queue.get(timeout=1.0)
+            uid = request_queue.get(timeout=0.01)
```

Comment:
Is there a reason why we set the timeout to be so low? It will unblock anyway when something shows up on the queue.

Reply:
yeah make sense. reverted to `1.0`